---
title: "Clustering Visualization and Analysis of Single Cell RNAseq data"
subtitle: "Sample(s): `r params$sample`, with dimensional reduction resolution set to `r params$resolution`."
author: "Paula A. Fernandes"
date: "`r format(Sys.Date(), '%d/%m/%Y')`"
params:
  sample: "Co-culture day 10"
  resolution: "0.8"
  inputdir: "/Users/paulafernandes/sc_projects_output/co-culture/wo_many_genes/co-cult_8pcas_clustering.RDS"
output: 
  html_document:
    theme: united
    highlight: tango
    toc: true
    toc_float:
      smooth_scroll: false
---

```{r}
knitr::opts_chunk$set(message = FALSE, echo = FALSE, warning = FALSE)
```

Pipeline based on Seurat library for single cell RNAseq.

```{r}
library(Seurat)
library(ggplot2)
library(gridExtra)
library(cowplot)
library(clustree)
library(dplyr)
library(ape)
library(metap)
library(kableExtra)
library(knitr)
library(readr)
library(DT)
library(limma)
library(readr)
library(slingshot)
library(MAST)
```

```{r}
inputdir = params$inputdir

print(inputdir)
```

```{r}
integrated_data = readRDS(inputdir)
```

**Total number of cells per condition**

```{r}
# parameters:
resolution = params$resolution

integrated_data
```



```{r}
table_cells = table(integrated_data@meta.data$orig.ident) 
table_cells = as.data.frame(table_cells)

colnames(table_cells) = c("Sample", "Cell Number")

table_cells %>% 
  kable() %>% 
  kable_styling(full_width = F)
```

# Dimensional Reduction

## How many principal components (PCs) to use?

PCs represent the primary sources of heterogeneity of the dataset.

**Genes present on the first 5 PCs**

```{r}
print(integrated_data[["pca"]], dims = 1:5, nfeatures = 5)
```

```{r, fig.width = 12, fig.height = 8}
DimHeatmap(integrated_data, dims = 1:5, cells = 500, balanced = TRUE)
```

The elbowplot below ranks the PCs based on the percentage its variance, suggesting that the majority of true signal is captured in the first PCs.

```{r}
ElbowPlot(integrated_data, ndims = 50, reduction = "pca") 
```

# Clustering

Step one: Seurat constructs of a KNN graph based on euclidean distance in PCA space - previously defined above.

Step two: Define clusters, which has an important parameter (resolution) that sets the "granularity" of the clusters. Higher resolution means higher number of clusters.

## Which resolution to use?

The FindClusters function implements the procedure, and contains a resolution parameter that sets the granularity of the downstream clustering, with increased values leading to a greater number of clusters.

```{r}
if(integrated_data@active.assay == "integrated") {
  plot_res0.2 =  DimPlot(integrated_data, reduction = "umap", group.by = "integrated_snn_res.0.2", label = TRUE) +
   ggtitle("Resolution 0.2")

  plot_res0.4 = DimPlot(integrated_data, reduction = "umap", group.by = "integrated_snn_res.0.4", label = TRUE) +
   ggtitle("Resolution 0.4")

  plot_res0.5 = DimPlot(integrated_data, reduction = "umap", group.by = "integrated_snn_res.0.5", label = TRUE) +
    ggtitle("Resolution 0.5")
  
   plot_res0.6 = DimPlot(integrated_data, reduction = "umap", group.by = "integrated_snn_res.0.6", label = TRUE) +
    ggtitle("Resolution 0.6")

   plot_res0.8 = DimPlot(integrated_data, reduction = "umap", group.by = "integrated_snn_res.0.8", label = TRUE) +
    ggtitle("Resolution 0.8")
   
   plot_res1.0 = DimPlot(integrated_data, reduction = "umap", group.by = "integrated_snn_res.1", label = TRUE) +
    ggtitle("Resolution 1")
   } else if(integrated_data@active.assay == "SCT") {
  plot_res0.2 =  DimPlot(integrated_data, reduction = "umap", group.by = "SCT_snn_res.0.2", label = TRUE) +
   ggtitle("Resolution 0.2")

  plot_res0.4 = DimPlot(integrated_data, reduction = "umap", group.by = "SCT_snn_res.0.4", label = TRUE) +
   ggtitle("Resolution 0.4")

  plot_res0.5 = DimPlot(integrated_data, reduction = "umap", group.by = "SCT_snn_res.0.5", label = TRUE) +
    ggtitle("Resolution 0.5")
  
   plot_res0.6 = DimPlot(integrated_data, reduction = "umap", group.by = "SCT_snn_res.0.6", label = TRUE) +
    ggtitle("Resolution 0.6")

   plot_res0.8 = DimPlot(integrated_data, reduction = "umap", group.by = "SCT_snn_res.0.8", label = TRUE) +
    ggtitle("Resolution 0.8")
   
   plot_res1.0 = DimPlot(integrated_data, reduction = "umap", group.by = "SCT_snn_res.1", label = TRUE) +
    ggtitle("Resolution 1")
   } else { 
     print("error: wrong active.assay")
  }
```

```{r, fig.width = 12, fig.height = 16}
   plot_grid(plot_res0.2, plot_res0.4, plot_res0.5, plot_res0.6, plot_res0.8, plot_res1.0,  nrow = 3, ncol = 2)
```

```{r}
# set the resolution
if (integrated_data@active.assay == "integrated") {
  Idents(integrated_data) = integrated_data@meta.data[[paste0("integrated_snn_res.", resolution)]]
} else if (integrated_data@active.assay == "SCT") {
  Idents(integrated_data) = integrated_data@meta.data[[paste0("SCT_snn_res.", resolution)]]
} else {
  print("error: wrong active.assay")
}
```

The clustree package below helps to visualize how cells are distributed between clusters depending on resolution

```{r, fig.width = 7, fig.height = 7}
if (integrated_data@active.assay == "integrated") {
  clustree(integrated_data@meta.data, prefix = "integrated_snn_res.")
} else if (integrated_data@active.assay == "SCT") {
  clustree(integrated_data@meta.data, prefix = "SCT_snn_res.")
} else {
  print("error: wrong active.assay")
}
```

```{r}
integrated_data = BuildClusterTree(integrated_data, dims = 1:30)
```

**ClusterTree**

```{r}
PlotClusterTree(integrated_data)
```

```{r, results = "asis"}
cat(paste0("#### For this dataset, set resolution to ", resolution), "\n")
```

# Non-linear dimensional reduction

UMAP and t-SNE are non-linear dimensional reduction techniques used to visualize and explore the datasets on a 2 dimensional plan. The goal is to place similar cells together, in a low dimensional space. We use the previously defined PC numbers as input for the clustering analysis.

## UMAP projection of the data

```{r}
group_by_plot = DimPlot(integrated_data, reduction = "umap", group.by = "orig.ident")
umap_plot = DimPlot(integrated_data, reduction = "umap", label = TRUE, repel = TRUE)
```

```{r, fig.width = 8, fig.height =7}
if (integrated_data@active.assay == "SCT") {
print(umap_plot) 
  } else {
  print(NULL)
}
```

Grouped by data time points

```{r, fig.width = 13, fig.height = 6}
if (integrated_data@active.assay == "integrated") {
print(group_by_plot + umap_plot) 
  } else {
    print(NULL)
}
```
Split by data time points

```{r, fig.width = 12, fig.height = 4}
if (integrated_data@active.assay == "integrated") {
  DimPlot(integrated_data, reduction = "umap", split.by = "orig.ident")
} else {
  print(NULL)
}
```

```{r}
total_num_cells_cluster = c()
if (integrated_data@active.assay == "integrated") {
  total_num_cells_cluster = as.data.frame(table(integrated_data@meta.data[[paste0("integrated_snn_res.", resolution)]]))
} else if (integrated_data@active.assay == "SCT") {
  total_num_cells_cluster = as.data.frame(table(integrated_data@meta.data[[paste0("SCT_snn_res.", resolution)]])) 
} else {
  print("error: wrong active.assay")
}
```

# Differentially Expressed Genes (DEG)

## Identify unique cell type markers for each cluster

Define DEGs based on the entire gene lists - not integrated one. Find unique markers for every cluster compared to all remaining cells. In the table below, it shows only the positive ones.

```{r}
integrated_data = PrepSCTFindMarkers(integrated_data)
DefaultAssay(integrated_data) <- "SCT"
```

```{r}
markers_all = FindAllMarkers(object=integrated_data, only.pos=TRUE)
```

```{r}
markers_all = markers_all[markers_all$p_val_adj < 0.05,]
markers_all_single = markers_all[markers_all$gene %in% names(table(markers_all$gene))[table(markers_all$gene) == 1],]
```

```{r}
gene_id_list = read_csv("/Users/paulafernandes/singelcell_rnaseq_seuratpipeline/gene_list/gene_id_list.csv")
  
markers_all_description = merge(x = markers_all, 
               y = gene_id_list, 
               by = "gene", 
               all.x = TRUE)
```

```{r}
gene_id_list = read_csv("/Users/paulafernandes/singelcell_rnaseq_seuratpipeline/gene_list/gene_id_list.csv")
  
markers_all_single_description = merge(x = markers_all_single, 
               y = gene_id_list, 
               by = "gene", 
               all.x = TRUE)
```

#### Unique markers per cluster: p-val \< 0.05

```{r}
datatable(markers_all_single_description, 
          extensions = 'Buttons', 
          options = list(dom = 'Bfrtip', 
                         buttons = c('copy', 'csv', 'excel', 'pdf'),
                         scrollX = TRUE)) 
```


```{r}
total_unq_markers_cluster = as.data.frame(table(markers_all_single$cluster))
colnames(total_unq_markers_cluster) = c("cluster", "Total Unique Markers")
```

```{r}
total_markers_cluster = as.data.frame(table(markers_all$cluster))
total_unq_markers_cluster["Total Markers"] = total_markers_cluster$Freq
```

```{r}
total_unq_markers_cluster["Total Number Cells"] = total_num_cells_cluster$Freq 
```

#### Total cell numbers per cluster

```{r}
total_unq_markers_cluster  %>%  
  kable() %>%
  kable_styling(full_width = F)
```

```{r}
top <- markers_all_single %>% group_by(cluster) %>% top_n(8, avg_log2FC)
integrated_data <- ScaleData(integrated_data)
```

#### Heatmap with the top unique genes per cluster, **based on log2 FC**

```{r, fig.height=12, fig.width=12}
DoHeatmap(object = integrated_data, features = top$gene)
```

## Identify **all** upregulated genes on each cluster, not necessarily unique

```{r}
markers_all_cluster = list()

for(cluster_number in sort(unique(markers_all_single_description$cluster))){
  markers_all_cluster[[cluster_number]] = filter(markers_all_single_description, cluster == cluster_number)
}
```

```{r, results="asis"}
clusterTableFeaturePlot = function(table){

  table = table[order(-table$avg_log2FC),]
  cat(paste0("### Cluster: ", unique(table$cluster), "\n"))
  cat("Based on log2 FC")
  cat("\n")
  
  print(table  %>%  
          kable() %>%
          kable_styling() %>%
          scroll_box(width = "100%", height = "300px"))
  
  cat(paste0("#### Markers for cluster: ", unique(table$cluster, "\n")))
  cat("\n")
  top_6 = table %>% group_by(cluster) %>% top_n(6, avg_log2FC)
  print(FeaturePlot(integrated_data, head(top_6$gene, n=6), cols = c("seashell2", "darkmagenta"), ncol = 2, order = TRUE))
  cat("\n")
  cat("\n")
  }
```

```{r, fig.width = 12, fig.height = 12, results = "asis"}
for(table in markers_all_cluster){
  clusterTableFeaturePlot(table)
  }
```

# Meiotic/gamete Markers

```{r}
meiotic_markers = as.matrix(c(Hap2 = "Tb927.10.10770", GEX1 = "Tb927.9.7040", MND1 = "Tb927.11.5670", DMC1 = "Tb927.9.9620", HOP1 = "Tb927.3.1750", SPO11 = "Tb927.5.3750"))
colnames(meiotic_markers) = "Gene IDs"

meiotic_markers  %>%  
  kable() %>%
  kable_styling(full_width = F)
```

```{r, fig.width = 12, fig.height = 12, results = "asis"}
meiotic_markers = c(Hap2 = "Tb927.10.10770",
                    GEX1 = "Tb927.9.7040",
                    MND1 = "Tb927.11.5670",
                    DMC1 = "Tb927.9.9620",
                    HOP1 = "Tb927.3.1750",
                    SPO11 = "Tb927.5.3750")

FeaturePlot(integrated_data, meiotic_markers, cols = c("seashell2", "darkmagenta"), ncol = 2, order = TRUE)
```

```{r, fig.width = 12, fig.height = 12, results = "asis"}
# replicate_sg_paper = c(Ep1 = "Tb927.10.10260", 
#                     Barp = "Tb927.9.15640", 
#                     VSG = "VSG-393",
#                     RBP6 = "Tb927.3.2930",
#                     Calflagin = "Tb927.8.5440", 
#                     Hap2 = "Tb927.10.10770")
# 
# FeaturePlot(integrated_data, replicate_sg_paper, cols = c("seashell2", "darkmagenta"), ncol = 2, order = TRUE, min.cutoff = 1)
```


```{r}
# cluster0.markers = FindMarkers(integrated_data, ident.1 = 0, only.pos= TRUE, test.use = "MAST", min.pct = 0.10)
# cluster1.markers = FindMarkers(integrated_data, ident.1 = 1, only.pos= TRUE, test.use = "MAST", min.pct = 0.10)
# cluster2.markers = FindMarkers(integrated_data, ident.1 = 2, only.pos= TRUE, test.use = "MAST",  min.pct = 0.10)
# cluster3.markers = FindMarkers(integrated_data, ident.1 = 3, only.pos= TRUE, test.use = "MAST", min.pct = 0.10)
# cluster4.markers = FindMarkers(integrated_data, ident.1 = 4, only.pos= TRUE, test.use = "MAST", min.pct = 0.10)
# cluster5.markers = FindMarkers(integrated_data, ident.1 = 5, only.pos= TRUE, test.use = "MAST", min.pct = 0.10)

```

```{r}
# markers_per_cluster = list()
# 
# for (cluster_num in 0:5){
#   markers_per_cluster[[cluster_num]] = FindMarkers(integrated_data, ident.1 = cluster_num, only.pos= TRUE, test.use = "MAST", min.pct = 0.10)
# }
```

```{r}
# cluster0.markers$gene = rownames(cluster0.markers)
# cluster1.markers$gene = rownames(cluster1.markers)
# cluster2.markers$gene = rownames(cluster2.markers)
# cluster3.markers$gene = rownames(cluster3.markers)
# cluster4.markers$gene = rownames(cluster4.markers)
# cluster5.markers$gene = rownames(cluster5.markers)
```

```{r}
# cluster0.markers = merge(x = cluster0.markers, 
#                y = gene_id_list, 
#                by = "gene", 
#                all.x = TRUE)
# 
# cluster1.markers = merge(x = cluster1.markers, 
#                y = gene_id_list, 
#                by = "gene", 
#                all.x = TRUE)
# 
# cluster2.markers = merge(x = cluster2.markers, 
#                y = gene_id_list, 
#                by = "gene", 
#                all.x = TRUE)
# 
# cluster3.markers = merge(x = cluster3.markers, 
#                y = gene_id_list, 
#                by = "gene", 
#                all.x = TRUE)
# 
# cluster4.markers = merge(x = cluster4.markers, 
#                y = gene_id_list, 
#                by = "gene", 
#                all.x = TRUE)
# 
# cluster5.markers = merge(x = cluster5.markers, 
#                y = gene_id_list, 
#                by = "gene", 
#                all.x = TRUE)
```

```{r}
#integrated_data$seurat_clusters = Idents(integrated_data)
#saveRDS(integrated_data, "/Users/paulafernandes/sc_projects_output/co-culture/wo_many_genes/integrated_clusters.RDS")
```

# Info Session

```{r}
sessionInfo()
```
